### ISSUE TITLE
Embedding model

model.py

### ISSUE THREAD

**Issue by captain-sal** [captain-sal/cybersecurity#1]
import torch.nn as nn

class LSTMModel(nn.Module):
    def __init__(self, input_dim, embedding_dim):
        super(LSTMModel, self).__init__()
        self.lstm = nn.LSTM(input_dim, 128, batch_first=True)
        self.fc = nn.Linear(128, embedding_dim)

    def forward(self, x):
        _, (h_n, _) = self.lstm(x)
        x = h_n[-1]
        x = self.fc(x)
        return x

def get_model(input_dim, embedding_dim):
    return LSTMModel(input_dim, embedding_dim)

tripletloss.py

import torch
import torch.nn as nn

class TripletLoss(nn.Module):
    def __init__(self, margin=1.0):
        super(TripletLoss, self).__init__()
        self.margin = margin

    def forward(self, anchor, positive, negative):
        positive_distance = torch.sum((anchor - positive) ** 2, axis=1)
        negative_distance = torch.sum((anchor - negative) ** 2, axis=1)
        loss = torch.mean(torch.maximum(positive_distance - negative_distance + self.margin, torch.zeros_like(positive_distance)))
        return loss

train.py

import torch
from torch.utils.data import DataLoader
from src.data_loader import load_data, TripletDataset
from src.model import get_model
from src.triplet_loss import TripletLoss

# Load data
x_train, x_test, y_train, y_test = load_data('data/raw/your_data.csv')

# Create datasets and loaders
train_dataset = TripletDataset(x_train, y_train)
test_dataset = TripletDataset(x_test, y_test)
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=32)

# Define model, loss, and optimizer
input_dim = x_train.shape[1]
embedding_dim = 64
model = get_model(input_dim, embedding_dim)
criterion = TripletLoss(margin=0.2)
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# Training and evaluation functions
def train(model, train_loader, optimizer, criterion, device):
    model.train()
    total_loss = 0
    for batch in train_loader:
        anchor, positive, negative = batch
        anchor, positive, negative = anchor.to(device), positive.to(device), negative.to(device)
        
        optimizer.zero_grad()
        
        anchor_output = model(anchor.unsqueeze(1))
        positive_output = model(positive.unsqueeze(1))
        negative_output = model(negative.unsqueeze(1))
        
        loss = criterion(anchor_output, positive_output, negative_output)
        loss.backward()
        optimizer.step()
        
        total_loss += loss.item()
    
    return total_loss / len(train_loader)

def evaluate(model, test_loader, criterion, device):
    model.eval()
    total_loss = 0
    with torch.no_grad():
        for batch in test_loader:
            anchor, positive, negative = batch
            anchor, positive, negative = anchor.to(device), positive.to(device), negative.to(device)
            
            anchor_output = model(anchor.unsqueeze(1))
            positive_output = model(positive.unsqueeze(1))
            negative_output = model(negative.unsqueeze(1))
            
            loss = criterion(anchor_output, positive_output, negative_output)
            total_loss += loss.item()
    
    return total_loss / len(test_loader)

# Train model
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)
num_epochs = 10
for epoch in range(num_epochs):
    train_loss = train(model, train_loader, optimizer, criterion, device)
    test_loss = evaluate(model, test_loader, criterion, device)
    print(f'Epoch {epoch + 1}, Train Loss: {train_loss:.4f}, Test Loss: {test_loss:.4f}')

# Save model
torch.save(model.state_dict(), 'outputs/models/lstm_model.pth')

inference.py

import torch
from torch.utils.data import DataLoader, TensorDataset
from src.model import get_model

def get_embeddings(model, data_loader, device):
    model.eval()
    embeddings = []
    with torch.no_grad():
        for batch in data_loader:
            data = batch[0].to(device)
            output = model(data.unsqueeze(1))
            embeddings.append(output.cpu().numpy())
    return np.concatenate(embeddings)

# Load test data
x_train, x_test, y_train, y_test = load_data('data/raw/your_data.csv')
test_dataset = TensorDataset(torch.FloatTensor(x_test))
test_loader = DataLoader(test_dataset, batch_size=32)

# Load model
input_dim = x_test.shape[1]
embedding_dim = 64
model = get_model(input_dim, embedding_dim)
model.load_state_dict(torch.load('outputs/models/lstm_model.pth'))
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model.to(device)

# Get embeddings
user_embeddings = get_embeddings(model, test_loader, device)
np.save('outputs/embeddings/user_embeddings.npy', user_embeddings)

requirements. txt

numpy
pandas
scikit-learn
torch
torchvision

Readme.md
# LSTM Model with Triplet Loss for User Profiling

This project implements an LSTM model trained with triplet loss to generate unique user profiles and embeddings from labeled data.

## Folder Structure

- `data/`: Contains raw and processed data.
- `src/`: Source code for data loading, model definition, training, and inference.
- `notebooks/`: Jupyter notebooks for data exploration
